javascript/react-js/react-testing-part-one.md

```diff
@@ -0,0 +1,177 @@
+
+We've learned testing our vanilla JavaScript applications in the previous section. Testing is indeed a powerful tool that allows to write maintainable and flexible code. If you've followed along our lessons, we've been using the [Jest](https://jestjs.io/) framework. For React, we'll keep using Jest and add more capabilities to our tests using the React Testing library.
+
+Read the [introduction](https://testing-library.com/docs/react-testing-library/intro) page of the library before we get started.
+
+### Learning Outcomes
+
+By the end of this lesson, you should be able to do the following:
+
+TODO
+
+### Setting Up
+
+We'll need to import some packages atop our test file:
+
+~~~javascript
+import React from "react"
+import { ... } from "@testing-library/react"
+import "@testing-library/jest-dom"  // optional
+import userEvent from "@testing-library/user-event" // optional
+import TestComponent from "path-to-test-component"
+~~~
+
+* `@testing-library/react` will give us access to useful functions like `render` which we'll demonstrate later on.
+
+* `@testing-library/jest-dom` includes some handy custom matchers (assertive functions) like `toBeInTheDocument` and more. (complete list [here](https://github.com/testing-library/jest-dom)). Jest already has a lot of matchers so this package is not compulsory to use. 
+
+* `@testing-library/user-event` provides the `userEvent` API that simulates user interactions with the webpage. It's optional because we can import the `fireEvent` API from the `@testing-library/react` instead; `userEvent` builds off of `fireEvent` to provide extra features. 
+
+* No need to import `jest` since it will automatically detect test files (`*.test.js` or `*.test.jsx`).
+
+That's a lot of setup. But good news! If you're initializing your React repositories with `create-react-app`, then all the above packages come preinstalled and the scripts preconfigured in `package.json`. 
+
+### Our First Query
+
+First, we'll render the component using `render`. The API will return an object and we'll use destructuring syntax to obtain a subset of the methods required. You can read all about what `render` can do [here](https://testing-library.com/docs/react-testing-library/api/#render).
+
+~~~javascript
+// App.js
+
+import React from "react"
+
+const App = () => <h1>Our First Test</h1>
+
+export default App
+~~~
+
+~~~javascript
+// App.test.js
+
+import React from "react"
+import { render } from "@testing-library/react"
+import App from "./App"
+
+test("correct heading renders", () => {
+    const { getByRole } = render(<App />)
+    expect(getByRole("heading").textContent).toMatch(/our first test/i)
+})
+~~~
+
+Execute `npm test App.js` on the terminal and you'll see that test pass. `getByRole` is just one of the dozen query methods that we could've used. Essentially, queries are classified into three types: `getBy`, `queryBy` and `findBy`. Read about their differences [here](https://testing-library.com/docs/queries/about/).  
+
```

Comment:
I think that it would be beneficial to touch on using the name option with getByRole, as in my experience people are often unaware of that functionality and that it's a "better" and "more RTL" approach to getting an element than using getByText or a test  ID.

Excerpt from [the RTL docs](https://testing-library.com/docs/queries/byrole/) advocating for the use of the name option with getByRole: 
![image](https://user-images.githubusercontent.com/7115053/149677279-a69ff7c0-7eea-4e0a-8097-d426e988f2db.png)

Reply:
I'd also like to see us explicitly talk about the RTL approach/ethos when it comes to [which queries should be favored over others](https://testing-library.com/docs/queries/about/#priority) in general